Rose user guide rose tutorial

oliver-sanders opened the pull request:
~**Built on #2166, hold review until merged (will resolve conflicts after this merge).**~

Rose tutorial for the new documentation:

* [x] What is rose
* [x] Rose applications
* [x] Rose metadata
* [x] Rose suite configurations
* [x] Rosie & FCM (needs finalisation)
* [x] Additional rose topics (built-in apps, rose utilities)

Main objectives for this content:

* Explain rose configurations [in a generic sense] (purpose, format).
* Convey the relationship between rose applications, suites and rose configurations.
* Clarify the purpose of rose applications (namely encapsulation).
* Clarify the purpose(s) of rose metadata.
* Clarify the relationship between rose suite configurations and cylc suites (cylc suites with rose configurations).
* Introduce and reinforce rose terminology.
* Introduce the tools for evaluating rose configurations
    * [x] `rose metadata-check`
    * [ ] ~`rose config`~
* Introduce the tools for running rose configurations
    * [x] `rose app-run`
    * [x] `rose suite-run`
    * [x] `rose task-run`

To do in a future PR:

* rose config?
* rose config edit section?
* rosie go section?
* rose stem section

~Ready for preliminary review (structure, approaches, etc).~

oliver-sanders commented:
This is now more or less finished - the Rosie section could do with a critical eye and possibly some contributions.

oliver-sanders commented:
Conflicts resolved. Ignore codacy issues regarding the use of `!important` in css files, it's not a nice thing to do but is required to reliably override theme values.

oliver-sanders commented:
Remaining test battery failure is pycode style which will be fixed when rebased on-top of #2172.

oliver-sanders commented on sphinx/cheat-sheet.rst:
```diff
@@ -0,0 +1,198 @@
+         # No special steps required.
+     - ::
+
+         # Only if the suite is not running:
```
The cylc reviewing commands (graph, validate, ...) can indeed operate on rose suites (irrespective of location). Unfortunately this falls apart when you define variables in the `rose-suite.conf` file, as cylc doesn't know about this file it coughs up an undefined variable error. Hopefully this will be solved along with the upcoming migration of `rose suite-run` functionality into cylc as it is a nasty un-necessary extra step.

sadielbartholomew replied:
I see. Thanks for explaining this.

oliver-sanders commented on sphinx/tutorial/rose/applications.rst:
```diff
@@ -0,0 +1,245 @@
+         # The weighting to give to the wind file from each WIND_CYCLE
+         # (should add up to 1).
+         WEIGHTING=1
+         # List of cycle points to get wind data from.
```
Updated comment explaining that its a comma separated list.

sadielbartholomew replied:
Thanks, I think it is a good idea.

matthewrmshin commented on sphinx/tutorial/rose/furthertopics/polling.rst:
```diff
@@ -26,7 +30,7 @@ file that looks like this:
                       bob => read_letter"""
 
-This sets up a simple suite which consists of the following:
+This is simple suite which consists of the following:
```
This is *a* simple ...

matthewrmshin commented on sphinx/tutorial/rose/rosie.rst:
```diff
@@ -0,0 +1,307 @@
+      Switch back to the trunk then merge your change branch into the trunk::
+
+         fcm sw trunk
+         fcm merge change trunk
```
`fcm merge change` should be sufficient.

matthewrmshin commented on sphinx/tutorial/rose/summary.rst:
```diff
@@ -0,0 +1,199 @@
+   See the :ref:`date-time tutorial <rose-tutorial-datetime-manipulation>`
+   for more information.
+
+:ref:`command-rose-host-select`
```
I think we should get rid of this one.

oliver-sanders replied:
After https://github.com/cylc/cylc/issues/1885 and https://github.com/cylc/cylc/issues/2199?

oliver-sanders commented on sphinx/tutorial/rose/summary.rst:
```diff
@@ -0,0 +1,199 @@
+
+The relationship between them is as follows:
+
+.. graph:: Example
```
Good point, for the moment I'll add diagrams / link the relevant sections through from the glossary.

oliver-sanders commented:
@matthewrmshin Note I've added a few new glossary items in 307639c that could probably do with a quick scan.

lhuggett commented:
Section: rose applications
The bit about creating a new directory called forecast. The directory
doesn't seem to be used in this section.
The rose-app.conf file: where is this supposed to be created? I put it in the
current directory (~/rose-tutorial/application-tutorial) which seems to work.
Viewing of map: works in firefox, but not in knoqueror.

Section: metadata
Having clicked on the link to the full list of metadata items,
it is quite confusing at this stage of the training.
Typing all the '[env=WHATEVER]' and 'description=some comment or other' stuff into
the rose-meta.conf file to set it up is a bit tedious, even when copying and pasting
most of it from the rose-app.conf file.

Section: rose suite configurations
I would suggest changing 'When we run rose applications in cylc suites we use the
rose task-run command rather than the rose app-run command.' to
When we run rose applications from within cylc suites we use the
rose task-run command rather than the rose app-run command.'
To avoid possible ambiguity.

The bit where it says 'TODO - remove triple quotes in cylc tutorial'
is that intended for the tuttee or the developer?

The changes to the forecast task - it doesn't say where to find this.
I have assumed in the suite.rc file.

With the setting of initial and final cycle points, it doesn't seem
to work with copying and pasting in the results from the cylc training.

The 'post_process_exeter' task currently fails due to not being able to
find file "+PT01H00M.csv"

Cheat sheet:
what is the difference between --path and -C?

matthewrmshin commented on sphinx/glossary.rst:
```diff
@@ -540,6 +546,111 @@ Glossary
+      * :term:`warm start`
+
+   warm start
+      In a :term:`datetime cycling` suite
```
Not necessarily *datetime*.

matthewrmshin commented on sphinx/glossary.rst:
```diff
@@ -540,6 +546,111 @@ Glossary
+      configuration.
+
+      Reloading changes to the ``[runtime]`` section is safe, reloading changes
+      to the ``[scheduling]`` section has caveats, see the `Cylc User Guide`_.
```
Note: `[runtime]` changes that affect the family inheritance hierarchy may affect scheduling as well.

oliver-sanders replied:
Oh, good point.

oliver-sanders replied:
See new commit.

oliver-sanders commented:
#### Section: rose applications

> The bit about creating a new directory called forecast. The directory
doesn't seem to be used in this section.

This should now be fixed.

> The rose-app.conf file: where is this supposed to be created? I put it in the
current directory (~/rose-tutorial/application-tutorial) which seems to work.

The tutorial now gets you to `cd application-tutorial` then create a `rose-app.conf` file so I hope this is now implicit, we could add a `touch rose-app.conf` step if it's still unclear.

> Viewing of map: works in firefox, but not in knoqueror.

Note sure why this is happening, I'll take a look at this one later.

#### Section: metadata

> Having clicked on the link to the full list of metadata items, it is quite confusing at this stage of the training.

Agreed, the API pages are much more technical. I'm not entirely sure what to do about this really. We could remove the link from the tutorial as it's not really necessary or update the text to read something like.

```
For future reference a full list of metadata items can be found ...
```

> Typing all the '[env=WHATEVER]' and 'description=some comment or other' stuff into
the rose-meta.conf file to set it up is a bit tedious, even when copying and pasting
most of it from the rose-app.conf file.

I was a bit worried about this. For the moment I'll put in a TODO for future work. Suggestions welcome.

#### Section: rose suite configurations

> I would suggest changing 'When we run rose applications in cylc suites we use the
rose task-run command rather than the rose app-run command.' to
When we run rose applications from within cylc suites we use the
rose task-run command rather than the rose app-run command.'
To avoid possible ambiguity.

Good point, done.

> The bit where it says 'TODO - remove triple quotes in cylc tutorial'
is that intended for the tuttee or the developer?

I've now moved this into a comment.

> The changes to the forecast task - it doesn't say where to find this.
I have assumed in the suite.rc file.

I've re-phrased this to note that the task is in the ``suite.rc`` file.

> With the setting of initial and final cycle points, it doesn't seem
to work with copying and pasting in the results from the cylc training.

Fixed.

> The 'post_process_exeter' task currently fails due to not being able to
find file "+PT01H00M.csv"

I'll need to see this error in context to advise, possibly a typo or the likes or a misleading instruction.

#### Cheat sheet:

> what is the difference between --path and -C?

One is correct and the other is a strange figment of my immigration. Well spotted, not quite sure where that came from, fixed.

sadielbartholomew commented on sphinx/tutorial/rose/suites.rst:
```diff
@@ -0,0 +1,611 @@
+            WIND_FILE_TEMPLATE=test-data/wind_{cycle}_{xy}.csv
+            RAINFALL_FILE=test-data/rainfall.csv
+            MAP_FILE=map.html
+            CYLC_TASK_CYCLE_POINT=20171101T0000Z
```
Noting for future improvements brief conversation in person regarding setting this variable to ``20171101T0000Z`` being a potential cause of confusion for readers, given it is a random date-time tied to the test data.

oliver-sanders replied:
I've inserted a better explanation of the test data in a42675a.

oliver-sanders commented on sphinx/tutorial/rose/rosie.rst:
```diff
@@ -0,0 +1,307 @@
+In Rosie suites the :term:`suite directory` is added to `version control`_
+using `FCM`_.
+
+FCM is a `subversion <SVN>`_ (SVN) wrapper which provides a standard working
```
Fixed.

sadielbartholomew commented on sphinx/glossary.rst:
```diff
@@ -540,6 +546,111 @@ Glossary
+      * :term:`Shutdown` and :term:`restarted <restart>`
+      * "Reloaded"
+
+      Reloading does not require the suite to be :term:`shutdown`. When a suite
```
``:term:`shut down <shutdown>` `` :)

sadielbartholomew commented on sphinx/glossary.rst:
```diff
@@ -540,6 +546,111 @@ Glossary
+
+      Reloading does not require the suite to be :term:`shutdown`. When a suite
+      is reloaded any currently "active" :term:`tasks <task>` will continue with
+      their "pre-reload" configuration, new tasks will use the new
```
'configuration, **whilst** new tasks...' perhaps?

sadielbartholomew commented on sphinx/tutorial/rose/applications.rst:
```diff
@@ -212,19 +222,13 @@ application:
       .. code-block:: rose
 
-         # The cycle point for the test data.
+         # The date when the test data was gathered.
```
Much clearer! There should be no confusion about this now.

sadielbartholomew commented on sphinx/glossary.rst:
```diff
@@ -643,8 +643,9 @@ Glossary
-      to the ``[scheduling]`` section has caveats, see the `Cylc User Guide`_.
+      Reloading changes is safe providing they don't affect the
+      :term:`suite's <suite>` :term:`graph`. Changes to the graph have certain
+      caveats attached, see the `Cylc User Guide`_.
```
attached. See the ... [for more information].' reads better.

sadielbartholomew commented on sphinx/installation.rst:
```diff
@@ -6,8 +6,8 @@ Installation
 
-The source code for Rose is available via `GitHub`_, code releases
-are available in ``zip`` and ``tar.gz`` `archives`_.
+The source code for Rose is available via `GitHub <Rose GitHub>`_, code
```
The link to ``SVN`` in ``sphinx/tutorial/rose/rosie.rst`` is also broken for the same reason.

oliver-sanders replied:
Doh, I keep walking into that one.

oliver-sanders commented on sphinx/tutorial/rose/applications.rst:
```diff
@@ -181,9 +181,9 @@ can be used to provide the path to the Rose application:
 
          # The weighting to give to the wind file from each WIND_CYCLE
-         # (should add up to 1).
+         # (comma separated list, values should add up to 1).
```
Done.

oliver-sanders commented on sphinx/glossary.rst:
```diff
@@ -540,6 +546,111 @@ Glossary
+
+   stop
+   shutdown
+      When a :term:`suite` is shutdown the :term:`suite server program` is
```
shutdown as in `cylc shutdown <suite>`

sadielbartholomew commented on sphinx/api/configuration/site-and-user.rst:
```diff
@@ -14,10 +14,18 @@ site configuration at ``$ROSE_HOME/etc/rose.conf`` where
 ``$HOME`` is the home directory of the current user.
 
-You can also override many internal constants of the ``rose config edit`` and
+.. note::
```
This note is no longer applicable, the ``rose.conf`` settings section having being added in now.

sadielbartholomew commented on sphinx/tutorial/cylc/scheduling/graphing.rst:
```diff
@@ -230,20 +230,18 @@ seen so far. The number ``1`` which appears below each task is the
          pub -> bar -> wop
 
-   #. **Use** ``cylc graph`` **to visualise the workflow.**
+   #. **Use** ``Cylc graph`` **to visualise the workflow.**
```
This is a command, so 'cylc' still needs to be lower-case here.

sadielbartholomew commented on sphinx/tutorial/cylc/scheduling/integer-cycling.rst:
```diff
@@ -406,15 +401,15 @@ starting 5 cycles *after* the initial cycle point.
    #. **Visualise the suite.**
 
-      Try visualising the suite using ``cylc graph``.
+      Try visualising the suite using ``Cylc graph``.
```
This is a command, so 'cylc' still needs to be lower-case here.

sadielbartholomew commented on sphinx/api/configuration/rose-configuration-format.rst:
```diff
@@ -13,9 +13,9 @@ following:
   INI format.
 
-We have added the following conventions into the rose configuraiton format:
+We have added the following conventions into the Rose configuraiton format:
```
I've noticed a typo ('configura**i**ton') - it's outside of the PR changes but we should get it fixed now.
